File: src/main/scala/com/wufuqiang/leetcode/scala/L2_AddTwoNumbers.scala
package com.wufuqiang.leetcode.scala

object L2_AddTwoNumbers {

  def compute(node1: ListNode, node2: ListNode ,result :ListNode):Unit={
    var tmp = node1.x + node2.x
    result.next = new ListNode(tmp % 10)
    if(node1.next == null && node2.next == null && tmp / 10 ==0){
      return
    }
    var node1Next = if(node1.next != null) node1.next else new ListNode(0)
    var node2Next = if(node2.next != null) node2.next else new ListNode(0)
    node1Next.x += tmp /10
    compute(node1Next,node2Next,result.next)
  }

  def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
    var result = new ListNode
    compute(l1,l2,result)
    result.next
  }
}


//   Definition for singly-linked list.
 class ListNode(var _x: Int = 0) {
   var next: ListNode = null
   var x: Int = _x
 }

File: src/main/java/com/wufuqiang/leetcode/L2_AddTwoNumbers.java
package com.wufuqiang.leetcode;

import com.wufuqiang.leetcode.entries.ListNode;

public class L2_AddTwoNumbers {
	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		ListNode result = new ListNode(0);

		int carry = 0;
		int sum = 0;
		ListNode cur = result;
		while(l1 != null && l2 != null){
			sum = l1.val+l2.val+carry;
			carry = sum / 10;
			l1.val = sum % 10;
			cur.next = l1;
			cur = l1;
			l1 = l1.next;
			l2 = l2.next;
		}
		while(l1 != null){
			sum = l1.val + carry;
			carry = sum / 10;
			l1.val = sum %10;
			cur.next = l1;
			cur = l1;
			l1 = l1.next;
		}

		while(l2 != null){
			sum = l2.val+carry;
			carry = sum /10;
			l2.val = sum %10;
			cur.next = l2;
			cur = l2;
			l2 = l2.next;
		}
		if(carry != 0){
			cur.next = new ListNode(carry);
		}

		return result.next;
	}
}
